lib/wibox/layout/ratio.lua

```diff
@@ -0,0 +1,339 @@
+-- @param context The context in which we are drawn.
+-- @param width The available width.
+-- @param height The available height.
+function ratio:layout(context, width, height)
+    local result = {}
+    local pos,spacing = 0, self._spacing
+    local num = #self.widgets
+    local total_spacing = (spacing*(num-1))
+
+    normalize(self)
+
+    for k, v in ipairs(self.widgets) do
+        local space = nil
+        local x, y, w, h
+
+        if self.dir == "y" then
+            space = height * self._ratios[k]
+            x, y = 0, util.round(pos)
+            w, h = width, floor(space)
+        else
+            space = width * self._ratios[k]
+            x, y = util.round(pos), 0
+            w, h = floor(space), height
+        end
+
+        table.insert(result, base.place_widget_at(v, x, y, w, h))
+
+        pos = pos + space + spacing
+
+        if (self.dir == "y" and pos-spacing >= height) or
+            (self.dir ~= "y" and pos-spacing >= width) then
+            break
+        end
+    end
+
+    return result
+end
+
+--- Increase the ratio of "widget"
+-- If the increment produce an invalid ratio (not between 0 and 1), the method
+-- do nothing.
+-- @tparam number index The widget index to change
+-- @tparam number increment An floating point value between -1 and 1 where the
+--   end result is within 0 and 1
+function ratio:inc_ratio(index, increment)
+    if #self.widgets ==  1 or (not index) or increment < -1 or increment > 1 then return end
+
+    local widget = self.widgets[index]
+
+    assert(self._ratios[index])
+
+    self:set_ratio(index, self._ratios[index] + increment)
+end
+
+--- Increment the ratio of the first instance of `widget`
+-- If the increment produce an invalid ratio (not between 0 and 1), the method
+-- do nothing.
+-- @param widget The widget to ajust
+-- @tparam number increment An floating point value between -1 and 1 where the
+--   end result is within 0 and 1
+function ratio:inc_widget_ratio(widget, increment)
+    if not widget or not increment then return end
+
+    local index = self:index(widget)
+
+    self:inc_ratio(index, increment)
+end
+
+--- Set the ratio of the widget at position `index`
+-- @tparam number index The widget to change or its index
+-- @tparam number percent An floating point value between 0 and 1
+function ratio:set_ratio(index, percent)
+    if not percent or #self.widgets ==  1 or not index or index > #self.widgets
+        or percent < 0 or percent > 1 then 
+        return
+    end
+
+    if not self._ratios[index] then
+        normalize(self)
+    end
+
+    local old = self._ratios[index]
+
+    -- Remove what has to be cleared from all widget
+    local delta = ( (percent-old) / (#self.widgets-1) )
+
+    for k, v in pairs(self.widgets) do
+        self._ratios[k] = self._ratios[k] - delta
+    end
+
+    -- Set the new ratio
+    self._ratios[index] = percent
+
+    -- As some widgets may now have a slightly negative ratio, normalize again
+    normalize(self)
```

Comment:
What's the idea for when to call `normalize()`? All functions seem to call `normalize()` before doing anything, so that suggests that the idea is "the widget may always be 'inconsistent'". However, by that reasoning, you wouldn't have to call `normalize` here.

However, for speed reasons (beware, micro-optimization!) I would suggest to change that rule to "the widget is always normalized". This means that nothing has to call `normalize(self)` before doing something, since the assumption is that we are already normalize. Only after changing things (e.g. here) does normalize have to be called. I guess that would mean that this is the only call to `normalize` that is left.